fmlcore/src/main/java/net/minecraftforge/fml/VersionChecker.java

```diff
@@ -116,46 +119,47 @@ public void run()
                     return;
                 }
 
+                client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(HTTP_TIMEOUT_SECS)).build();
                 gatherMods().forEach(this::process);
             }
 
             /**
-             * Opens stream for given URL while following redirects
+             * Returns the response body as a String for the given URL while following redirects
              */
-            private InputStream openUrlStream(URL url) throws IOException
-            {
+            private String openUrlString(URL url) throws IOException, URISyntaxException, InterruptedException {
                 URL currentUrl = url;
                 for (int redirects = 0; redirects < MAX_HTTP_REDIRECTS; redirects++)
                 {
-                    URLConnection c = currentUrl.openConnection();
-                    if (c instanceof HttpURLConnection)
+                    var request = HttpRequest.newBuilder()
+                            .uri(currentUrl.toURI())
+                            .timeout(Duration.ofSeconds(HTTP_TIMEOUT_SECS))
+                            .GET()
+                            .build();
+
+                    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+
+                    int responseCode = response.statusCode();
+                    if (responseCode >= 300 && responseCode <= 399)
                     {
-                        HttpURLConnection huc = (HttpURLConnection) c;
-                        huc.setInstanceFollowRedirects(false);
-                        int responseCode = huc.getResponseCode();
-                        if (responseCode >= 300 && responseCode <= 399)
+                        try
                         {
-                            try
-                            {
-                                String loc = huc.getHeaderField("Location");
-                                currentUrl = new URL(currentUrl, loc);
-                                continue;
-                            }
-                            finally
-                            {
-                                huc.disconnect();
-                            }
+                            String newLocation = response.headers().map().get("Location").get(0);
+                            currentUrl = new URL(currentUrl, newLocation);
+                            continue;
+                        }
+                        catch (NullPointerException e)
+                        {
+                            throw new IOException("Got a 3xx response code but Location header was null while trying to fetch " + url);
```

Comment:
This should use a null check, not a NPE catch.
Or really this should just be
`String newLocation = response.headers().firstValue("Location").orElseThrow(...)`

Reply:
`firstValue("Location")` cannot be used as the returned response might not always have `Location` as the first header.

I've implemented an alternative approach that takes this into account but still uses `Optional`, thanks for your feedback.

Reply:
Headers might be repeated. `firstValue` simply returns the first Location header, not the first header should it be the `Location`. `firstValue` uses code equivalent to what you have currently.

Reply:
My mistake, I misunderstood the JavaDoc. Tested it now and you're right, using `firstValue()` instead...